packages/server/src/workers/longterm-job.ts

```diff
@@ -0,0 +1,136 @@
+  }
+
+  async updateStatus(job: Job<TData>, output: Parameters): Promise<void> {
+    const exec = new AsyncJobExecutor(this.systemRepo, job.data.asyncJob);
+    const updatedJob = await exec.updateJobProgress(this.systemRepo, output, {
+      // Conditional update to ensure this update does not clobber another,
+      // which could result in e.g. continuing a job that was cancelled
+      ifMatch: job.data.asyncJob.meta?.versionId,
+    });
+
+    if (updatedJob) {
+      job.data.asyncJob = updatedJob;
+    }
+  }
+
+  async finishJob(job: Job<TData>, output?: Parameters): Promise<void> {
+    const exec = new AsyncJobExecutor(this.systemRepo, job.data.asyncJob);
+    await exec.completeJob(this.systemRepo, output);
+  }
+
+  async failJob(job: Job<TData>, err?: Error): Promise<void> {
+    const exec = new AsyncJobExecutor(this.systemRepo, job.data.asyncJob);
+    await exec.failJob(this.systemRepo, err);
+  }
+
+  async checkJobStatus(job: Job<TData>): Promise<boolean> {
+    const asyncJob = await this.systemRepo.readResource<AsyncJob>('AsyncJob', job.data.asyncJob.id as string);
+
+    if (!inProgressJobStatus.includes(asyncJob.status)) {
+      return false;
+    }
+
+    job.data.asyncJob = asyncJob;
+    return true;
+  }
+
+  async execute(job: Job<TData>): Promise<void> {
+    const canStart = await this.checkJobStatus(job);
+    if (!canStart) {
+      // Job is not in-progress, terminate early
+      return;
+    }
+
+    try {
+      const result = await this.process(job);
+
+      // Check if AsyncJob resource should be updated; this usually
+      // happens less frequently than every iteration
+      const output = this.formatResults(result, job);
+      if (output) {
+        try {
+          await this.updateStatus(job, output);
+        } catch (err) {
+          if (err instanceof OperationOutcomeError && getStatus(err.outcome) === 412) {
+            // Conflict: AsyncJob was updated by another party between when the job started and now!
+            // Check status to see if job was cancelled
+            const canContinue = await this.checkJobStatus(job);
+            if (!canContinue) {
+              // Job was cancelled or errored in parallel; this iteration should abort
+              return;
+            }
+          }
+
+          throw err;
+        }
+      }
+
+      const nextIteration = this.nextIterationData(result, job);
+      if (typeof nextIteration === 'boolean') {
+        // Job is complete, no more iterations should be enqueued
+        const jobSucceeded = nextIteration;
+        if (jobSucceeded) {
+          await this.finishJob(job, output);
+        } else {
+          await this.failJob(job);
+        }
+      } else {
+        // Enqueue job for the specified next iteration
+        // NOTE: We do not check the AsyncJob status before enqueuing: it will be checked
+        // at the beginning of the next iteration
+        await this.enqueueJob(nextIteration);
+      }
+    } catch (err: any) {
```

Comment:
Small nit:

Prefer `err: unknown`

With case to a specific error type in the function arguments